### PR TITLE
Include Cachi2's name in the SBOM metadata

### PR DESCRIPTION
**Pull request by brunoapimentel**
Cachi2 is included in the .metadata.tools section:
```
{
  ...,
  "metadata": {
    "timestamp": "2023-07-11T15:29:36.730637",
    "tools": [
      {
        "name": "Cachi2",
        "vendor": "Red Hat"
      }
    ]
  },
  "specVersion": "1.4",
  "version": 1
}
```

And also in each component's properties:

```
{
  "name": "bytes",
  "properties": [
    {
      "name": "cachi2:foundBy",
      "value": "cachi2"
    }
  ],
  "purl": "pkg:golang/bytes?type=package",
  "type": "library"
}
```

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

File: cachi2/core/models/output.py [hunk 1]
```diff
@@ -45,6 +47,12 @@ def from_package_dict(cls, package: dict[str, Any]) -> "Component":
         )
 
 
+class SbomMetadata(pydantic.BaseModel):
+    """Metadata field in a SBOM."""
+    timestamp: datetime = pydantic.Field(default_factory=datetime.utcnow)
```
Comment (brunoapimentel) [inline cachi2/core/models/output.py:52]: Pydantic states that `default_factory` is in [beta](https://docs.pydantic.dev/1.10/usage/models/#field-with-dynamic-default-value), but the alternative was to override the __init__ function, which felt worse than this. Thoughts?

Comment (brunoapimentel) [reply cachi2/core/models/output.py:52]: Btw, I added this `timestamp` field because I thought it was nice to have, but I can remove it altogether if you don't value in it.

Comment (brunoapimentel) [reply cachi2/core/models/output.py]: Nevermind, I got rid of the timestamp once I realized we'd need to change the integration tests to handle it properly :sweat_smile:

File: cachi2/core/models/output.py [hunk 2]
```diff
@@ -19,6 +19,7 @@ class Component(pydantic.BaseModel):
 
     name: str
     version: Optional[str]
+    properties: list[dict[str, str]] = [{"name": "cachi2:foundBy", "value": "cachi2"}]
```
Comment (chmeliik) [inline cachi2/core/models/output.py:22]: 1. can we have a Property model?
2. do we want our properties to be camelCase?
3. when we want to add a non-static property from the package manager code, we will override this default and lose the foundBy

File: cachi2/core/models/output.py [hunk 3]
```diff
@@ -45,6 +46,12 @@ def from_package_dict(cls, package: dict[str, Any]) -> "Component":
         )
 
 
+class SbomMetadata(pydantic.BaseModel):
+    """Metadata field in a SBOM."""
+
+    tools: list[dict[str, str]] = [{"vendor": "Red Hat", "name": "Cachi2"}]
```
Comment (chmeliik) [inline cachi2/core/models/output.py:52]: Can we have a Tool model?

(using the CycloneDX library is looking more and more tempting TBH)

File: cachi2/core/models/output.py (hunk 2 above)
Comment (chmeliik) [reply cachi2/core/models/output.py:22]: Maybe the name should also be capitalized consistently, it's `cachi2` here but `Cachi2` in tools

TBH I prefer all lowercase, but it's already Cachi2 everywhere

Comment (chmeliik) [reply cachi2/core/models/output.py:22]: Although maybe we can use `cachi2` in code and Cachi2 in documentation

Comment (brunoapimentel) [reply cachi2/core/models/output.py:22]: I think we can keep as lowercase "cachi2" at least within the SBOM.

Comment (brunoapimentel) [reply cachi2/core/models/output.py]: 2. I copied that from Syft without a second thought. I think the spec suggests it to use hyphens: https://github.com/CycloneDX/cyclonedx-property-taxonomy

Comment (brunoapimentel) [reply cachi2/core/models/output.py]: 3. Not sure how to improve this. Maybe we can add this "found-by" to all components as part of the RequestOutput creation, wdyt?

File: cachi2/core/models/output.py (hunk 3 above)
Comment (brunoapimentel) [reply cachi2/core/models/output.py]: I started a sample commit with the library, but it proved to be more troublesome than I expected. File a story to analyse this later?

File: cachi2/core/models/output.py (hunk 2 above)
Comment (chmeliik) [reply cachi2/core/models/output.py]: I don't think the spec suggests any naming style, even the `cdx` namespace itself is not consistent (mostly [camelCase](https://github.com/CycloneDX/cyclonedx-property-taxonomy/blob/main/cdx/device.md#cdxdevice-namespace-taxonomy), some kebab-case ([1](https://github.com/CycloneDX/cyclonedx-property-taxonomy/blob/main/cdx/device.md#cdxdevicegs1-namespace-taxonomy), [2](https://github.com/CycloneDX/cyclonedx-property-taxonomy/blob/main/cdx/npm.md#cdxnpmpackageconstraint-namespace-taxonomy)), sometimes literal [env variable names](https://github.com/CycloneDX/cyclonedx-property-taxonomy/blob/main/cdx/gomod.md#cdxgomodbuildenv-namespace-taxonomy))

I would choose something and stick to it, personally prefer snake_case

Comment (chmeliik) [reply cachi2/core/models/output.py]: > we can add this "found-by" to all components as part of the RequestOutput creation, wdyt?

depends on what you mean by RequestOutput creation :smile: 

One way could be a properties `@validator` that appends the found_by

Comment (brunoapimentel) [reply cachi2/core/models/output.py]: > I would choose something and stick to it, personally prefer snake_case
I don't mind snake_case. Unless someone else has a different opinion, let's stick to that.

**Review by ben-alkov — APPROVED** [review]
LGTM

File: cachi2/core/models/output.py [hunk 4]
```diff
@@ -32,6 +43,13 @@ def key(self) -> str:
 
         return f"{self.name}:{self.version}"
 
+    @pydantic.validator("properties", always=True)
+    def _add_found_by_property(cls, properties: list[Property]) -> list[Property]:
+        if FOUND_BY_CACHI2_PROPERTY not in properties:
+            properties.append(FOUND_BY_CACHI2_PROPERTY)
+
+        return properties
```
Comment (chmeliik) [inline cachi2/core/models/output.py:51]: IMO we can append the property unconditionally. I don't see us accidentally adding the `FOUND_BY_CACHI2_PROPERTY` from package manager code ever :smile:

**Review by chmeliik — APPROVED** [review]
LGTM with a minor nitpick :+1:

File: cachi2/core/models/output.py (hunk 4 above)
Comment (chmeliik) [reply cachi2/core/models/output.py:51]: Unless that duplicates them when merging the SBOMs of different package managers https://github.com/containerbuildsystem/cachi2/blob/03ee6507b7e06bcd1393ff826982e6d95a89bbd7/cachi2/core/resolver.py#L35

**Review by lkolacek — APPROVED** [review]
LGTM :+1:

File: cachi2/core/models/output.py (hunk 4 above)
Comment (brunoapimentel) [reply cachi2/core/models/output.py:51]: It seems that the validation is only triggered a the component is instantiated. Adding it to another model, even in the constructor, does not trigger a validation (at least, it didn't in the test I did).

I would probably just leave the condition there, so we could avoid a corner case.
